odd-platform-api/src/main/java/com/provectus/oddplatform/repository/DatasetStructureRepositoryImpl.java

```diff
@@ -0,0 +1,76 @@
+package com.provectus.oddplatform.repository;
+
+import com.provectus.oddplatform.model.tables.pojos.DatasetFieldPojo;
+import com.provectus.oddplatform.model.tables.pojos.DatasetStructurePojo;
+import com.provectus.oddplatform.model.tables.pojos.DatasetVersionPojo;
+import com.provectus.oddplatform.model.tables.records.DatasetStructureRecord;
+import lombok.RequiredArgsConstructor;
+import org.jooq.DSLContext;
+import org.jooq.InsertSetStep;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static com.provectus.oddplatform.model.Tables.DATASET_STRUCTURE;
+
+@Repository
+@RequiredArgsConstructor
+public class DatasetStructureRepositoryImpl implements DatasetStructureRepository {
+    private final DSLContext dslContext;
+    private final DatasetFieldRepository datasetFieldRepository;
+    private final DatasetVersionRepository datasetVersionRepository;
+
+    @Override
+    @Transactional
+    public List<DatasetStructurePojo> bulkCreate(final List<DatasetVersionPojo> versions,
+                                                 final Map<Long, List<DatasetFieldPojo>> datasetFields) {
+        final List<DatasetFieldPojo> fields = datasetFields.values().stream()
+            .flatMap(List::stream)
+            .collect(Collectors.toList());
+
+        final List<DatasetVersionPojo> createdVersions = datasetVersionRepository.bulkCreate(versions);
```

Comment:
What if some version already exists? I haven't looked deeply but I don't see any protection mechanism.
On the next line there's a call `bulkCreateIfNotExist` and on this one we don't have a special treatment for existing records. Is this handled somehow?

Reply:
There's no API for updating dataset_versions, we either create or read them. A protection mechanism lays a bit higher on the stack, but I see your point and agree that we should make a single approach for Ingestion API.